### ISSUE TITLE
Noise

### ISSUE THREAD

**Issue by StuartIanNaylor** [linto-ai/linto-desktoptools-hmg#8]
I was reading https://doc.linto.ai/#/client/custom_hotword as have been wondering how often noise is misconstrued and how to incorporate it into a model.

> "You’ll need 2 kinds of audio samples:
> 
> A bunch of recording of the word you want to use.
> A bunch of recording of random speech/noise audio where the word is not pronounce."

Also you can gaussian noise in the training section but to be honest I am not sure if or how that is applied.
The above has confused me as to make the model resilient to noise you need noise with the KW not the !KW samples otherwise you will be reducing resultant confidence score.
A common method seems to mix noise & KW at various SNR of 5dB, 10dB, 15dB that duplicates clean KW and spilts into three ratios.

The whole noise thing can be confusing and tedious and wondering if again further options to defines a noise folder, ratio steps, step level and % of KWs to duplicate.
Also maybe here as the file locations are known pitch shift + temp, pad shifting automation to fill out a KW set by slight modification of the duplicates.

Maybe its just wording but slightly confused by what is said in the documentation and that noise is always confusing and why options to employ a standard process in the HMG would be a good idea.

**Comment by StuartIanNaylor** [comment]
PS I have been looking at other networks than a GRU even though a GRU works fine I am under the impression tensorflow for microcontrollers doesn't yet support?

A CRNN for size & latency looks the best but not much infront of a GRU and still the RNN part that at least doesn't seem easy to transfer onto microcontrollers.

Do you know if the current model can be exported to TensorFlow Lite for Microcontrollers as really interested in distributed wide array KWS with cheap microcontrollers and unidirectional electrets where best confidence hit of a zone gives current working audio stream.

**Comment by Lokhozt** [comment]
Hello Stuart,
I have indeed in mind to had a noising module to add noise to existing samples. The "random speech/noise" designation might be a little confusing i reckon, and doesn't mean "noise" as in "noise over signal" but rather "audio that is not speech". There are 2 schools for handling noised signal: The first being to train on clean samples then remove it with preprocessing and the second is to train the model to be noise resilient using noised samples during training. I found that the MFCC / GRU implementation is not particularly good for the task. There are some tweaks used to improve signal separation and more adapted features for the task.

PS: Once the release is done i'll conduct some test on CRNN and add it in HMG if the test are conclusive. As for TF for microcontrollers i have to admit that it not being a target platform for LinTO at the moment, i have no knowledge in that matter. It is interesting though, i'll look into it when i have some spare time.

**Comment by StuartIanNaylor** [comment]
I can see you working away so just thinking of any items that might be worth considering now before your code is set in stone :)

I am itching to give the HMG a go and will do some experiments with noise with the first method which I find seems to work OK where I duplicate my KW split in quarters norm and then 5db, 10db, 15db, 20db split 25, 25, 25,25% as it also adds to my tally of KW which helps balance !KW which is usually the google command set.

The pre-processing for a satellite for me needs hardware of too much$ but its a different consideration for a shared ASR stream that maybe might have accelerated realtime models.

I am still thinking in terms of wide array distributed KWS where best confidence level picks the mic sic stream for that specific KW session.
There is a ESP32 with a AC101 codec built in AI Thinker A1S where pre-processing isn't an option but for KW I wouldn't use it but would for a central ASR as that GPU can be shared and also provide ASR to TTS.

Google are going crazy on streaming KWS and they have a whole section of practically every model you could think of in  
https://github.com/google-research/google-research/tree/master/kws_streaming
https://arxiv.org/abs/2005.06720

A new one on me is  Singular value decomposition filter (SVDF) even lighter than CRNN.

> "In Table 2 we observe that the most effective and accurate streaming models are SVDF, CRNN and GRU."

Table 2 in the above is worth a look

**Comment by StuartIanNaylor** [comment]
PS can I ask a question asmy MS amnesia has totally forgotten.

What method of the model favours false negatives over false positives as remember it was set to 0.9 and completely forgot what the method was without reinstalling.

**Comment by StuartIanNaylor** [comment]
PS https://github.com/tranHieuDev23/TC-ResNet Looks interesting and that is a Keras implementation of https://github.com/hyperconnect/TC-ResNet

**Comment by StuartIanNaylor** [comment]
@Lokhozt are we there yet? :) lols the kid in the backseat asks, but wondering purely for testing shall I give it a go.

PS I still disagree about what you said about noise as what you said was true but I just have a hunch the hardware that HMG will be used to create models for lacks the mips to do noise reduction as also I have played with noise reduction and its only been high end 'RTXVoice' like frameworks that are of any use.
Lower down the scale the audio artefacts outweigh the SNR reduction and removing noise from no voice segments is pointless anyway.

ASR really benefits from the pooling of hardware that may service distributed KWS and noise reduction is more possible but for KWS its cost prohibitive and also not nessacary with singular known and trained KW.
I think what you say is correct but actually when it comes to application a preferential option is losing out to one that for KW is not likely to garner much adoption.

**Comment by StuartIanNaylor** [comment]
@Lokhozt its stayed static for a couple of days is it beta ready for some rough testing?

**Comment by Lokhozt** [comment]
Yes it is. By the way, I took a look into kws_streaming and there is a lot of interesting elements. CRNN and noise management will be in the next update.

**Comment by StuartIanNaylor** [comment]
I will be running an instance later :) thnx

The CRNN looks a great low ops NN its a shame about TF4MC but still a great choice.

Noise I would prob just set up some percentage splits and SNR as KW and Noise need to be normalised as otherwise you are cluless of resultant SNR.
I am thinking percent splits and dB level via 3 splits as I would prob choose 33% 5dB, 10dB, 15dB but maybe it should be choice.
Then maybe you just inch across a set of noise files so that no repetition creates anything distinct and also a tick box to enable/disable noise addition.

@Lokhozt PS a DS-CNN is documented in kws_streaming more ops but does run on TF4MC https://github.com/42io/esp32_kws
